### PR TITLE
KAFKA-7895: Fix stream-time reckoning for suppress

### PR DESCRIPTION
**Pull request by vvcephei**
* Add suppress to system tests
* Move stream-time reckoning from Task into Processor

Even within a Task, different Processors have different perceptions
of time, due to record caching on stores and in suppression itself,
and in general, due to any processor logic that may hold onto
records arbitrarily and emit them later. Thanks to this, we can't rely
on the whole task existing in the same "instant" of stream-time. The
solution is for each processor node that cares about stream-time to
track it independently.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
@ableegoldman @mjsax @bbejeck @guozhangwang ,

Please take a look at this fix for suppression.

As I mentioned in the description, it was incorrect for a processor to use the task's view of stream time, when a processor within a task may have an arbitrarily delayed local view of time, when it has an upstream record cache or suppression.

This was actually always true (since I introduced stream time into the context), but it didn't manifest as a bug until suppression, which depends very strictly on the stream time.

The correct thing to do is track observed stream time locally in each processor node. Since not every processor needs stream time, and since it was never in the ProcessorContext public API, I opted to remove it from the ProcessorContext and just explicitly implement it in the nodes that need it.

If we wish to expose "stream time" to processors in general, we can instead track stream time in ProcessorNode and then wrap the ProcessorContext to supply the local stream time when injecting the context to `Processor.init`. But I that can wait until the need arises. The current change is sufficient for now.

Ideally, we can merge this fix into 2.2 as well.

**Comment by vvcephei** [comment]
I guess it should be noted that the same condition could cause segments to be prematurely expired and records to be prematurely rejected from windowed aggregations due to grace period restrictions, if the windowed aggregation were downstream of a cache or suppression.

It just happens that suppression was the first thing that people noticed (probably due to the long default retentions and grace period).

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java [hunk 1]
```diff
@@ -81,6 +81,7 @@ public void enableSendingOldValues() {
         private StreamsMetricsImpl metrics;
         private InternalProcessorContext internalProcessorContext;
         private Sensor lateRecordDropSensor;
+        private long observedStreamTime = -1L;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: nit: use `ConsumerRecord.NO_TIMESTAMP` in stead of `-1L` (same below)

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java [hunk 2]
```diff
@@ -75,8 +75,6 @@ public void close() {}
         }
     };
 
-    private long streamTime = RecordQueue.UNKNOWN;
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java:78]: Can you fix `RecordQueue.UNKNOWN` as a side thing, too?

File: streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractSegments.java [hunk 3]
```diff
@@ -72,8 +72,8 @@ public S getSegmentForTimestamp(final long timestamp) {
     }
 
     @Override
-    public S getOrCreateSegmentIfLive(final long segmentId, final InternalProcessorContext context) {
-        final long minLiveTimestamp = context.streamTime() - retentionPeriod;
+    public S getOrCreateSegmentIfLive(final long segmentId, final InternalProcessorContext context, final long streamTime) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/AbstractSegments.java]: nit: break parameters into single lines

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java [hunk 4]
```diff
@@ -205,7 +209,9 @@ void restoreAllInternal(final Collection<KeyValue<byte[], byte[]>> records) {
             for (final Map.Entry<KeyValueSegment, WriteBatch> entry : writeBatchMap.entrySet()) {
                 final KeyValueSegment segment = entry.getKey();
                 final WriteBatch batch = entry.getValue();
-                segment.write(batch);
+                if (segment.open) {
+                    segment.write(batch);
+                }
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Can you elaborate? Seems to be orthogonal to the timestamp fix.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java [hunk 5]
```diff
@@ -414,8 +417,11 @@ public void shouldLogAndMeasureExpiredRecords() {
         LogCaptureAppender.setClassLoggerToDebug(RocksDBSegmentedBytesStore.class);
         final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
 
-        context.setStreamTime(Math.max(retention, segmentInterval) * 2);
-        bytesStore.put(serializeKey(new Windowed<>("a", windows[0])), serializeValue(5));
+        bytesStore.put(serializeKey(new Windowed<>("dummy", nextSegmentWindow)), serializeValue(0));
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java:430]: Is this to "set" a specific `obserservedTime` -- might need a comment

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java [hunk 6]
```diff
@@ -414,8 +417,11 @@ public void shouldLogAndMeasureExpiredRecords() {
         LogCaptureAppender.setClassLoggerToDebug(RocksDBSegmentedBytesStore.class);
         final LogCaptureAppender appender = LogCaptureAppender.createAndRegister();
 
-        context.setStreamTime(Math.max(retention, segmentInterval) * 2);
```
Comment (mjsax) [inline streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java:417]: I can't map this to the `nextSegmentWindow` bounds above. Can you elaborate?

**Comment by vvcephei** [comment]
Java 11 failure was a kafka admin test. Apparently, gradle crashed in the Java 8 build!

Retest this, please.

**Comment by mjsax** [comment]
https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/2402/
```
java.lang.AssertionError: The consumer group deletion did not work as expected
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.assertTrue(Assert.java:42)
	at kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdWithShortInitialization(DeleteConsumerGroupsTest.scala:211)
```
Java8 timed out (cf https://issues.apache.org/jira/browse/KAFKA-7933

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java (hunk 4 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: The segmented store expires segments based on stream time.

Previously, the stream time it used was coming from the ProcessorContext, which doesn't get updated while we're processing this batch write, so the whole method execution happened in one "instant". That instant would be at least the latest time in the batch. So when it loaded all the segments in building `batchWriteMap`, it was already able to tell which records were for expired segments and drop them.

Now, though, it tracks stream time based on the sequence of received writes, so at the beginning of this method call, the local stream time is maybe at the beginning of the batch. This means that in the beginning of this method, it creates/gets segments for everything in the batch (not updating the stream time at all), so all the segments are still live until we enter the loop over `writeBatchMap`.

When we loop over the batch map, we're actually processing the writes, which advances the local view of stream time, possibly expiring some segments. Since stream-time is actually advancing during the loop, and segments are being expired as the loop executes, we can no longer expect all our segments to be open here.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java (hunk 5 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java:430]: yes, it's to set the stream time to a high enough value that it expires the previous segment... I'll add a comment.

File: streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java (hunk 6 above)
Comment (vvcephei) [reply streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStoreTest.java:417]: The new code is a tighter bound. Previously we used `2 * max(retention, interval)`, but all we really need is `retention + interval`. Not sure why I did the previous code, it's guaranteed to be at least `retention + interval`, but it's pretty esoteric...

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: Hmm, should we use Long.MIN_VALUE instead, to account for negative timestamps?

We don't need a sentinel value, we just need something that's not larger than a real record's timestamp.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java (hunk 2 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java:78]: You mean, set it to `ConsumerRecord.NO_TIMESTAMP`, right?

**Comment by vvcephei** [comment]
Thanks for the review, @mjsax . I think I've addressed your comments.

**Review by bbejeck — COMMENTED** [review]
Thanks for the patch @vvcephei.  I've made one pass and I have one question regarding the test.  The smoke test covers the scenario of a time window for a 24 hour period with a grace of one minute, then sends records way beyond the time bounds to force a flush.  

That makes sense, but can we also test with tighter time bounds with windows of say 30 seconds and use system time without adding any extra time for verification?

**Comment by vvcephei** [comment]
I think that's a good idea, @bbejeck .

I'm thinking this could be an integration test, to avoid adding too much on to the smoke test, do you agree?

Actually, after wrestling with this for a bit, it's more complicated than I thought. What do you think about tackling that in a separate PR? I went ahead and added it to the acceptance criteria of KAFKA-7895. There's also a task there to add system test coverage with caching disabled, so this PR wouldn't have resolved the ticket anyway.

**Comment by vvcephei** [comment]
Failures unrelated:
https://builds.apache.org/job/kafka-pr-jdk8-scala2.11/19515/
```
kafka.api.ConsumerBounceTest.testRollingBrokerRestartsWithSmallerMaxGroupSizeConfigDisruptsBigGroup
org.apache.kafka.common.errors.GroupMaxSizeReachedException: Consumer group group2 already has the configured maximum number of members.
```

https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/2414/
```
kafka.api.PlaintextConsumerTest.testConsumeMessagesWithCreateTime

java.lang.AssertionError: Partition [__consumer_offsets,0] metadata not propagated after 15000 ms

```

Retest this, please,

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java (hunk 1 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: So far, we don't support negative timestamp, so it seems to be cleanest to use `NO_TIMESTAMP` for now -- if we ever introduce negative timestamps, we need to update the code anyway.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/StandbyContextImpl.java:78]: Yes.

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java (hunk 4 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Thanks for the details. Should we instead do a "clean up" within `getWriteBatches()` and only return open segments? Seems to be a better separation of concerns.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java (hunk 1 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamSessionWindowAggregate.java]: Sounds good. It's updated.

**Review by mjsax — APPROVED** [review]
LGTM (assuming Jenkins passes)

File: streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java (hunk 4 above)
Comment (vvcephei) [reply streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBSegmentedBytesStore.java]: Hah. Ok, that's what I get for commenting from memory. It turns out that we actually do advance time within `getWriteBatches`. The problem is actually that in that method, we're advancing time while we're also creating the segments and populating the result list. So, by the end of the iteration, some of the segments at the beginning of the result list may already be expired.

We can fix it by either pre-iterating over all the records to advance time to the largest timestamp in the batch or post-iterating over the results to remove the expired segments.

I opted for the pre-iteration so that we can potentially avoid creating and opening a segment that would just need to be closed and cleaned up later in the same batch.

The code is cleaned up now. Thanks for the catch!

**Comment by vvcephei** [comment]
Java 11 had one unrelated failure:

**Comment by vvcephei** [comment]
Java 11 had one unrelated failure:
https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/2424/
```
kafka.api.PlaintextConsumerTest.testMultiConsumerRoundRobinAssignment
java.lang.AssertionError: Did not get valid initial assignment for partitions [topic2-3, topic1-2, topic2-0, topic2-5, topic1-4, topic2-7, topic2-1, topic2-6, topic1-0, topic1-3, topic2-4, topic1-1, topic2-2]
	at kafka.utils.TestUtils$.fail(TestUtils.scala:356)
	at kafka.utils.TestUtils$.waitUntilTrue(TestUtils.scala:766)
	at kafka.api.PlaintextConsumerTest.validateGroupAssignment(PlaintextConsumerTest.scala:1772)
	at kafka.api.PlaintextConsumerTest.testMultiConsumerRoundRobinAssignment(PlaintextConsumerTest.scala:864)
```

**Comment by vvcephei** [comment]
That plaintext  consumer test was the only failure. I'm not sure if there's value in repeating the tests at this point.

**Comment by mjsax** [comment]
https://builds.apache.org/job/kafka-pr-jdk11-scala2.12/2424/
```
java.lang.AssertionError: Did not get valid initial assignment for partitions [topic2-3, topic1-2, topic2-0, topic2-5, topic1-4, topic2-7, topic2-1, topic2-6, topic1-0, topic1-3, topic2-4, topic1-1, topic2-2]
	at kafka.utils.TestUtils$.fail(TestUtils.scala:356)
	at kafka.utils.TestUtils$.waitUntilTrue(TestUtils.scala:766)
	at kafka.api.PlaintextConsumerTest.validateGroupAssignment(PlaintextConsumerTest.scala:1772)
	at kafka.api.PlaintextConsumerTest.testMultiConsumerRoundRobinAssignment(PlaintextConsumerTest.scala:864)
```
Retest this please.

**Comment by vvcephei** [comment]
Hey, it passed that time! Thanks, @mjsax .

**Comment by bbejeck** [comment]
>I'm thinking this could be an integration test, to avoid adding too much on to the smoke test, do you agree?

Yes, I actually was thinking of an integration test and just wasn't being clear about it.  

>Actually, after wrestling with this for a bit, it's more complicated than I thought. What do you think about tackling that in a separate PR? I went ahead and added it to the acceptance criteria of KAFKA-7895. There's also a task there to add system test coverage with caching disabled, so this PR wouldn't have resolved the ticket anyway.

That's fine.

**Review by bbejeck — COMMENTED** [review]
Thanks for the patch @vvcephei.  Over all looks good I just wanted to ask a clarifying question.  

Would it be fair to say the new approach "smooths" out the representation of time per processor, in that it can only ever increase or at a minimum stay the same?  

The process of only updating the `observedStreamTime` when the `internalProcessorContext#timestamp()` is greater than the last observed time is more or less the same approach taken in the `RecordQueue#maybeUpdateTimestamp()` correct?

**Comment by mbragg02** [comment]
Looks good, thanks @vvcephei

**Comment by vvcephei** [comment]
Thanks, @bbejeck ,

Yes, for the processors that track stream time, the definition of time is essentially the same as it is for the record queue. It's always the maximum record timestamp ever observed, so, as you say, it will always increase or stay the same.

It used to be true that (when there's one or more upstream caches or suppressions in the same task), we'd always see records arriving from the past. E.g., the clock would say 10:00, but I'd be getting records from 09:55. Now, we've made stream time _tighter_, so the clock will say 09:55 when we're getting records for 09:55.

Another change is that, for the processors tracking stream time, time only advances for records that actually flow through the processor, so we won't see the clock continuing to "tick" as a result of records that actually get filtered out before they get to us. I think this is good for predictability. That is, it's a simpler mental model, so it's easier to predict how the windowed aggs and suppression will behave wrt stream time.

I think it makes sense to say that this smoothes out the representation of stream time.

**Comment by vvcephei** [comment]
Thanks @mbragg02!

**Review by bbejeck — APPROVED** [review]
Thanks @vvcephei LGTM

**Comment by mjsax** [comment]
@vvcephei This PR has conflicts with `2.2` branch and I could not cherry-pick it cleanly. Can you please do a follow up PR against `2.2` for this? Thanks.

**Comment by mjsax** [comment]
Merged to `trunk`.

**Comment by vvcephei** [comment]
Thanks, @mjsax ! I will do so now.
